src/components/ec/cuda/ec_cuda_executor_interruptible.c

```diff
@@ -0,0 +1,157 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2020-2022.  ALL RIGHTS RESERVED.
+ *
+ * See file LICENSE for terms.
+ */
+
+#include "ec_cuda_executor.h"
+#include "components/mc/ucc_mc.h"
+
+ucc_status_t ucc_cuda_executor_interruptible_get_stream(cudaStream_t *stream)
+{
+    static int last_used = 0;
+    cudaError_t cuda_st;
+    int i, j;
+
+    ucc_spin_lock(&ucc_ec_cuda.init_spinlock);
```

Comment:
why we need lock always? do we need to check the thread mode?
Can we implement this "get_stream" using ucc_mpool_get? E.g. we can allocate a pool of ptrs with "init" function that actually creates a stream. Then it would be called when required. wdyt?

Reply:
I intentionally didn't use mpool because the idea here is to start reuse streams when we reach exec_num_streams. Alternatively we can use mpool, but then in some cases it will grow allocating a lot of streams, I wanted to avoid this. Do you think it's better not to limit number of streams?
cc @bureddy

Reply:
right, this makes sense. the question regarding locking still remains: we should probably use atomic_fadd for last_used in MT case and just regular "++" in ST. ?

Reply:
with mpool api have a parameter to set max limit.

Reply:
@bureddy yes it has "max_limit" but not clear how to re-use round robin way. we don't want a stream being taken from pool until the op is done. we want them to be used round-robin no matter how many outstanding ops are there. E.g., with 8 ops being posted and 4 streams max, each would get 2 ops. If we use mpool, then a stream taken from there is not available until it is returned back.

Reply:
@vspetrov, added atomic add instead of spinlock for main path. For stream init spinlock is still used, i think it's ok because we do it only once and lock/unlock overhead is small compared to stream create, also all ec/mc components initialized with UCC_THREAD_MULTIPLE and I don't think this will be fixed soon